examples/experimental/imagenet_nas_arch_pytorch/data.py

```diff
@@ -0,0 +1,118 @@
+    client = storage.Client()
+    for split in ['validation', 'train']:
+        prefix = os.path.join('imagenet/imagenet', split)
+        blobs = client.list_blobs(client, 'determined-ai-datasets', prefix=prefix)
+        for b in blobs:
+            blob.download_to_filename(os.path.join(data_dir, b.name))
+
+class ImageNetDataset(Dataset):
+    def __init__(self, split, bucket_name, transform=None):
+        """
+        Args:
+            split: train or validation split to return the right dataset
+            directory: root directory for imagenet where "train" and "validation" folders reside
+        """
+        assert split in ["train", "validation"], "split {} not in (train, validation)".format(split)
+        self._split = split
+        self._bucket_name = bucket_name
+        self._target_dir = '/mnt/data'
+        self._source_dir = os.path.join('imagenet/imagenet', self._split)
+        self._transform = transform
+        if self._bucket_name is not None:
+            self._storage_client = storage.Client()
+            self._bucket = self._storage_client.bucket(bucket_name)
+            # When the dataset is first initialized, we'll loop through to catalogue the classes (subdirectories)
+            # This step might take a long time.
+            self._imgs_paths = list_blobs(self._storage_client, self._bucket_name, prefix=self._source_dir)
+
+            self._labels = []
+            self._subdir_to_class = {}
+            class_count = 0
+            for path in self._imgs_paths:
+                sub_dir = path.split("/")[-2]
+                if sub_dir not in self._subdir_to_class:
+                    self._subdir_to_class[sub_dir] = class_count
+                    class_count += 1
+                self._labels.append(self._subdir_to_class[sub_dir])
+            print("There are {} records in dataset.".format(len(self._imgs_paths)))
+            self._streaming = True
+
+    def __len__(self):
+        # Return some length if using randomly generated data.
+        if self._bucket_name is None:
+            return 1024*100
+        return len(self._imgs_paths)
+
+
+    def __getitem__(self, idx):
+        if self._bucket_name is None:
+            array = np.random.rand(256, 256, 3)
+            return Image.fromarray(array, mode='RGB'), np.random.choice(1000)
+        img_path = self._imgs_paths[idx]
+        blob = self._bucket.blob(img_path)
+        if self._streaming:
+            img_str = download_gcs_blob_with_backoff(blob)
+        else:
+            target_path = os.path.join(self._target_dir, img_path)
+            if not os.path.exists(target_path):
+                create_file_dirs(target_path)
+                print('downloading...')
+                img_str = download_gcs_blob_with_backoff(blob)
+                with open(target_path,'wb') as f:
+                    f.write(img_str)
+            else:
+                with open(target_path, 'rb') as f:
+                    img_str = f.read()
+
```

Comment:
lets add that there is an option to pick stream and download or just download. This can get a bit confusing here.